### PR TITLE
feat(dgraph): making all internal communications with tls configured

### PR DESCRIPTION
**Pull request by aman-bansal**
Fixes DGRAPH-2505, DGRAPH-2506 and DGRAPH-546. All internal communication can now be TLS encrypted.
Three flags have been exposed tls_internal_port_enabled, tls_cert and tls_key to enable internal communication. 

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6692)

 
 

Docs Preview: [Dgraph Preview](https://dgraph-24f59d0b67-103439.surge.sh)

### CODE REVIEW

File: dgraph/cmd/bulk/run.go [hunk 1]
```diff
@@ -89,6 +89,8 @@ func init() {
 	flag.Bool("version", false, "Prints the version of Dgraph Bulk Loader.")
 	flag.Bool("store_xids", false, "Generate an xid edge for each node.")
 	flag.StringP("zero", "z", "localhost:5080", "gRPC address for Dgraph zero")
+	flag.String("zero_tls_dir", "",
```
Comment (parasssh) [inline dgraph/cmd/bulk/run.go:92]: ISnt this name confusing? This dir contains the bulk loader client's certs, no?

File: dgraph/cmd/zero/run.go [hunk 2]
```diff
@@ -104,23 +109,31 @@ type state struct {
...
+		grpc.StatsHandler(&ocgrpc.ServerHandler{}),
+	}
+
+	tlsConf, err := x.GenerateServerTLSConfig(x.LoadInternalTLSServerHelperConfig(Zero.Conf.GetString("dgraph_tls_dir")))
+	x.Check(err)
+
+	if tlsConf != nil {
+		grpcOpts = append(grpcOpts, grpc.Creds(credentials.NewTLS(tlsConf)))
+	}
+	s := grpc.NewServer(grpcOpts...)
```
Comment (parasssh) [inline dgraph/cmd/zero/run.go:125]: Will this change with c-mux in your next PR?

File: dgraph/cmd/zero/zero.go [hunk 3]
```diff
@@ -226,18 +228,23 @@ func (s *Server) SetMembershipState(state *pb.MembershipState) {
 	if state.Zeros == nil {
 		state.Zeros = make(map[uint64]*pb.Member)
 	}
+
```
Comment (parasssh) [inline dgraph/cmd/zero/zero.go:231]: remove vertical spacing.

File: dgraph/cmd/zero/zero.go [hunk 4]
```diff
@@ -226,18 +228,23 @@ func (s *Server) SetMembershipState(state *pb.MembershipState) {
...
 	if state.Groups == nil {
 		state.Groups = make(map[uint32]*pb.Group)
 	}
+
+	//todo check if these groups are all zero or if there is any alpha
```
Comment (parasssh) [inline dgraph/cmd/zero/zero.go:236]: how do you mean? 
Group 0 - all zero
Group 1..n - all alphas.

File: x/tls_helper.go [hunk 5]
```diff
@@ -55,9 +55,55 @@ func RegisterClientTLSFlags(flag *pflag.FlagSet) {
 		"provided by the client to the server.")
 }
 
+func RegisterDgraphTLSFlags(flag *pflag.FlagSet) {
+	flag.String("dgraph_tls_dir", "",
+		"Path to directory that has mTLS certificates and keys for dgraph internal communication")
+	flag.String("dgraph_tls_client_name", "",
+		"client name to be used for mTLS for dgraph internal communication")
+	flag.String("dgraph_tls_server_name", "",
```
Comment (parasssh) [inline x/tls_helper.go:63]: we dont need to differentiate client and server names. They can be the same thing.

File: x/tls_helper.go [hunk 6]
```diff
@@ -207,13 +253,40 @@ func GenerateServerTLSConfig(config *TLSHelperConfig) (tlsCfg *tls.Config, err e
 	return nil, nil
 }
 
-// GenerateClientTLSConfig creates and returns a new client side *tls.Config with the
+// GenerateClientTLSConfig creates and returns a new client side *tls.Config with tzhe
```
Comment (parasssh) [inline x/tls_helper.go:256]: spelling ... tzhe

File: dgraph/cmd/zero/zero.go (hunk 4 above)
Comment (aman-bansal) [reply dgraph/cmd/zero/zero.go:236]: Yes this was for my understanding to see which connections are made. I will remove these

File: dgraph/cmd/zero/run.go (hunk 2 above)
Comment (aman-bansal) [reply dgraph/cmd/zero/run.go:125]: No this won't change. cmux will only control zero HTTP port.

**Review by aman-bansal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6692)* status: 79 of 103 files reviewed, 11 unresolved discussions (waiting on @manishrjain, @parasssh, @pawanrawal, and @vvbalaji-dgraph)

___
*[conn/pool.go, line 149 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKAj8-M7-hFKJegWGs6:-MKKhlpy5RQD4CZj-1_C:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4c1867118f161511dba960d5bd227adb62f938f5/conn/pool.go#L149)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Can generate it at startup and then pass it here.nn
</blockquote></details>

Done.
___
*[dgraph/cmd/bulk/run.go, line 92 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MJDkzpCV1_U7BsjfKBY-r1-92:-MKKhoWu6K6x5yO18mTO:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/a5dd8a8eea2ead3454d85b02d341c1fabf055495/dgraph/cmd/bulk/run.go#L92)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

ISnt this name confusing? This dir contains the bulk loader client's certs, no?
</blockquote></details>

Done.
___
*[dgraph/cmd/live/run.go, line 547 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKAjn8dCMDwJWuMdUOy:-MKKhqHf7xlsoQvOul9o:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4c1867118f161511dba960d5bd227adb62f938f5/dgraph/cmd/live/run.go#L547)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

these 4 lines can go in one func and be reused everywhere
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/raft.go, line 216 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKAk9kV0RaVmOVMfou_:-MKKhrNl7u2GwXtMF57d:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4c1867118f161511dba960d5bd227adb62f938f5/dgraph/cmd/zero/raft.go#L216)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

can this be part of the `node` struct
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/run.go, line 125 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MJDkzpCV1_U7BsjfKB_-r1-125:-MKKhupL2uLncmWQrtO3:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/a5dd8a8eea2ead3454d85b02d341c1fabf055495/dgraph/cmd/zero/run.go#L125)):*
<details><summary><i>Previously, aman-bansal (aman bansal) wrote…</i></summary><blockquote>

No this won't change. cmux will only control zero HTTP port. 
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/run.go, line 56 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKAkK5_9jfiOD9nVGrL:-MKKhtmfB5_YXCskvu33:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4c1867118f161511dba960d5bd227adb62f938f5/dgraph/cmd/zero/run.go#L56)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

`// tls...`
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/zero.go, line 231 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MJDkzpCV1_U7BsjfKBa-r1-231:-MKKhvuS2Zl3E_dnEaAf:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/a5dd8a8eea2ead3454d85b02d341c1fabf055495/dgraph/cmd/zero/zero.go#L231)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

remove vertical spacing. 
</blockquote></details>

Done.
___
*[dgraph/cmd/zero/zero.go, line 236 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MJDkzpCV1_U7BsjfKBa-r1-236:-MKKhwWG2V-ApQhfTeFt:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/a5dd8a8eea2ead3454d85b02d341c1fabf055495/dgraph/cmd/zero/zero.go#L236)):*
<details><summary><i>Previously, aman-bansal (aman bansal) wrote…</i></summary><blockquote>

Yes this was for my understanding to see which connections are made. I will remove these
</blockquote></details>

Done.
___
*[x/tls_helper.go, line 63 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MJDkzpDZLHKFEpuKthz-r1-63:-MKKi3dWDSAk0xMK6sba:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/a5dd8a8eea2ead3454d85b02d341c1fabf055495/x/tls_helper.go#L63)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

we dont need to differentiate client and server names. They can be the same thing. 
</blockquote></details>

Done.
___
*[x/tls_helper.go, line 256 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MJDkzpDZLHKFEpuKthz-r1-256:-MKKi4da4yL4YqPmfLQH:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/a5dd8a8eea2ead3454d85b02d341c1fabf055495/x/tls_helper.go#L256)):*
<details><summary><i>Previously, parasssh wrote…</i></summary><blockquote>

spelling ... tzhe
</blockquote></details>

Done.
___
*[tlstest/mtls_internal/multi_group/mtls_internal_multi_group_test.go, line 36 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKAldOP6MPIK0k2TIfS:-MKKi-QlE1O53Lc5FIaT:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4c1867118f161511dba960d5bd227adb62f938f5/tlstest/mtls_internal/multi_group/mtls_internal_multi_group_test.go#L36)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

send mutation to one alpha and query the other two alphas
</blockquote></details>

Done.

**Review by aman-bansal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6692)* status: 61 of 111 files reviewed, 13 unresolved discussions (waiting on @manishrjain, @parasssh, @pawanrawal, and @vvbalaji-dgraph)

___
*[tlstest/mtls_internal/acl/docker-compose.yml, line 48 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKhk6Xn66vE2wMGr2LV:-MKhk6Xn66vE2wMGr2LW:b-e16kw7) ([raw file](https://github.com/dgraph-io/dgraph/blob/af116c081cf2ed14c8585bbd25565f5b600b063c/tlstest/mtls_internal/acl/docker-compose.yml#L48)):*
> ```
> --tls_dir /dgraph-tls
> ```
___
*[tlstest/mtls_internal/loader/loader_test.go, line 74 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKhkj0CBxMgYG9Q0sDT:-MKhkj0CBxMgYG9Q0sDU:b-anpj00) ([raw file](https://github.com/dgraph-io/dgraph/blob/af116c081cf2ed14c8585bbd25565f5b600b063c/tlstest/mtls_internal/loader/loader_test.go#L74)):*
> ```
> "--tls_enable_inter_node=true"
> ```
tls_internal_port_enabled

**Review by pawanrawal — APPROVED** [review]
:lgtm:  with some comments

> Reviewed 12 of 105 files at r2, 47 of 51 files at r3, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-:-MKhrAGuBmePpTPbVEoL:b-d09tk9)* status: all files reviewed, 11 unresolved discussions (waiting on @aman-bansal, @manishrjain, @parasssh, and @vvbalaji-dgraph)

___
*[dgraph/cmd/alpha/run.go, line 171 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKefx6z1zDwA9y8tImD:-MKefx7-0aqDbFk-bwqs:binp941) ([raw file](https://github.com/dgraph-io/dgraph/blob/ec511547478d37b2eb4371264f2b03031d3ebde3/dgraph/cmd/alpha/run.go#L171)):*
> ```Go
> 	flag.String("tls_client_auth", "VERIFYIFGIVEN", "Enable TLS client authentication")
> 	flag.Bool("tls_enable_inter_node", false, "(optional) enable inter node TLS encryption between cluster nodes.")
> 	flag.String("tls_cert", "", "(optional) The Cert file provided for the nodes to connect with the cluster.")
> ```

connect as a client with other nodes in the cluster.

Similarly below
___
*[dgraph/cmd/alpha/run.go, line 671 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKegFP22l1vCyX-Zbo9:-MKegFP385KBCawXGM96:b-6k63hy) ([raw file](https://github.com/dgraph-io/dgraph/blob/ec511547478d37b2eb4371264f2b03031d3ebde3/dgraph/cmd/alpha/run.go#L671)):*
> ```Go
> 		TLSClientConfig:      tlsConf,
> 		TLSDir:               Alpha.Conf.GetString("tls_dir"),
> 		TLSInterNodeEnabled: Alpha.Conf.GetBool("tls_enable_inter_node"),
> ```

Is `gofmt` run?
___
*[tlstest/mtls_internal/ha_6_node/ha_test.go, line 138 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKhqP-w6ORy8C1-Mb8S:-MKhqP-w6ORy8C1-Mb8T:b-lzpl6a) ([raw file](https://github.com/dgraph-io/dgraph/blob/af116c081cf2ed14c8585bbd25565f5b600b063c/tlstest/mtls_internal/ha_6_node/ha_test.go#L138)):*
> ```Go
> 	dgConn, err := grpc.Dial(":" + port, grpc.WithTransportCredentials(credentials.NewTLS(tlsConf)))
> 	require.NoError(t, err)
> 	time.Sleep(time.Second * 6)
> ```

Why is this sleep required?

**Review by aman-bansal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6692)* status: 98 of 111 files reviewed, 9 unresolved discussions (waiting on @manishrjain, @parasssh, @pawanrawal, and @vvbalaji-dgraph)

___
*[dgraph/cmd/alpha/run.go, line 171 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKefx6z1zDwA9y8tImD:-MKj4RMN7iU690bBoU5Q:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/ec511547478d37b2eb4371264f2b03031d3ebde3/dgraph/cmd/alpha/run.go#L171)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

connect as a client with other nodes in the cluster.

Similarly below
</blockquote></details>

Done.
___
*[dgraph/cmd/alpha/run.go, line 671 at r3](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKegFP22l1vCyX-Zbo9:-MKj4TQa7oLYxUSHyjVd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/ec511547478d37b2eb4371264f2b03031d3ebde3/dgraph/cmd/alpha/run.go#L671)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Is `gofmt` run?
</blockquote></details>

Done.
___
*[tlstest/mtls_internal/ha_6_node/ha_test.go, line 138 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6692#-MKhqP-w6ORy8C1-Mb8S:-MKj4YIn48gYifWuX1K3:bx1dasy) ([raw file](https://github.com/dgraph-io/dgraph/blob/af116c081cf2ed14c8585bbd25565f5b600b063c/tlstest/mtls_internal/ha_6_node/ha_test.go#L138)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

Why is this sleep required?
</blockquote></details>

this is Done. Actually I write the test case in different way initially that requires wait for the cluster to setup. changed it to docker compose setup later. I missed this refactoring.

**Pull request merged**